File: src/csharp/Intel/Iced/Intel/FormatterInternal/FormatterString.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if GAS || INTEL || MASM || NASM || FAST_FMT
using System.Diagnostics;

namespace Iced.Intel.FormatterInternal {
	readonly struct FormatterString {
		readonly string lower;
#if GAS || INTEL || MASM || NASM
		readonly string upper;
#endif

		public int Length => lower.Length;

		public FormatterString(string lower) {
			Debug.Assert(lower.ToLowerInvariant() == lower);
			this.lower = lower;
#if GAS || INTEL || MASM || NASM
			upper = string.Intern(lower.ToUpperInvariant());
#endif
		}

		public static FormatterString[] Create(string[] strings) {
			var res = new FormatterString[strings.Length];
			for (int i = 0; i < strings.Length; i++)
				res[i] = new FormatterString(strings[i]);
			return res;
		}

#if GAS || INTEL || MASM || NASM
		public string Get(bool upper) =>
			upper ? this.upper : lower;
#endif

#if FAST_FMT
		public string Lower => lower;
#endif
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/internal/fmt/FormatterString.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.internal.fmt;

import java.util.Locale;

/** DO NOT USE: INTERNAL API */
public final class FormatterString {
	private final String lower;
	private final String upper;

	public int getLength() {
		return lower.length();
	}

	public FormatterString(String lower) {
		assert lower.toLowerCase(Locale.ROOT) == lower : lower;
		this.lower = lower;
		upper = lower.toUpperCase(Locale.ROOT).intern();
	}

	public static FormatterString[] create(String[] strings) {
		FormatterString[] res = new FormatterString[strings.length];
		for (int i = 0; i < strings.length; i++)
			res[i] = new FormatterString(strings[i]);
		return res;
	}

	public String get(boolean upper) {
		return upper ? this.upper : lower;
	}

	public String getLower() {
		return lower;
	}
}
